Add a basic faq feature for rooms

bumbadadabum opened the pull request.

panpawn commented on chat-plugins/faq.js:
```diff
@@ -0,0 +1,59 @@
+		saveRoomFaqs();
+		this.privateModCommand(`(${user.name} has removed the faq for '${topic}'.)`);
+	},
+	faq: function (target, room, user) {
```
.... this is already a command...?

bumbadadabum replied:
NGPJRPWGMP SHIT

bumbadadabum replied:
I'm retarded don't mind me

panpawn commented:
Why not just save these to the room object...?

bumbadadabum commented:
Mostly because of clutter. I prefer to not potentially save 100 strings in chatRoomData.

panpawn commented:
I don't think 100 FAQ commands would be a thing for one room...

Maybe have a limit?

panpawn commented:
I was thinking to combine this with the regular faq command, is that a bad idea?

panpawn commented on chat-plugins/faq.js:
```diff
@@ -0,0 +1,59 @@
+let roomFaqs = {};
+try {
+	roomFaqs = require(ROOMFAQ_FILE);
+} catch (e) {} // file doesn't exist or contains invalid JSON
```
silent crashes aren't favorable

Zarel replied:
panpawn has a point here :p

bumbadadabum replied:
This is copied from another chat plugin.

Morfent replied:
Odd, there didn't seem to be a problem with it when Zarel helped me write it https://github.com/Zarel/Pokemon-Showdown/pull/1239#issuecomment-57916279
Regardless, this code for loading the json doean't crash, but it doesn't handle situations like the user who launched PS lacking the right privileges to edit the file, or the file already being used by another process. If you want to do it the best way, by all means fix it rather than arguing over precedent.

Zarel replied:
Apparently I just didn't notice that you took it out without doing it properly. I actually told you how to do it properly in the PR comments: `e indexof SyntaxError`. :|

Zarel replied:
Although doing it the other way is probably better:

```
} catch (e) {
  if (e.code !== 'MODULE_NOT_FOUND') throw e;
}
```

panpawn commented:
Maybe this should be a namespace command.

panpawn commented on chat-plugins/faq.js:
```diff
@@ -0,0 +1,59 @@
+			  "/roomfaq <topic> - Shows the faq for <topic>.",
+			  "/addfaq <topic>, <text> - Adds an entry for <topic> in this room or updates it. Requires: # & ~",
+			  "/removefaq <topic> - Removes the entry for <topic> in this room. Requires: # & ~"],
+};
```
isn't it a thing that we put empty lines on the bottom of files?

bumbadadabum commented:
Namespaces won't work here because there's no way to differentiate the way of getting the roomfaq for "add" and adding a roomfaq topic.

panpawn commented:
I don't think a room would put an FAQ entry for 'add'.

bumbadadabum commented:
the thing is it won't work with '': function() because of it. "/faq foo" would be interpreted as the command 'foo'.

panpawn commented on chat-plugins/faq.js:
```diff
@@ -0,0 +1,59 @@
+		if (!topic) return this.sendReplyBox("List of topics in this room: " + Object.keys(roomFaqs[room.id]));
+		if (!this.runBroadcast()) return;
+
+		if (!roomFaqs[room.id][topic]) return this.sendReply("Invalid topic.");
```
I feel like this should be an ``errorReply`` and also more specific.

panpawn commented on chat-plugins/faq.js:
```diff
@@ -0,0 +1,59 @@
+		if (!this.runBroadcast()) return;
+
+		if (!roomFaqs[room.id][topic]) return this.sendReply("Invalid topic.");
+		return this.sendReply(roomFaqs[room.id][topic]);
```
Why not a ``sendReplyBox`` to mirror the display of the regular faq command?

bumbadadabum commented on chat-plugins/faq.js:
```diff
@@ -0,0 +1,59 @@
+	roomfaq: function (target, room, user) {
+		if (!roomFaqs[room.id]) return this.errorReply("This room has no faq topics.");
+		let topic = toId(target);
+		if (!topic) return this.sendReplyBox("List of topics in this room: " + Object.keys(roomFaqs[room.id]));
```
Oh yeah I forgot to add that actually.

panpawn commented:
I feel like this should be disabled for battle rooms, and maybe groupchats, too.

panpawn commented:
There's no point in having rooms that are meant to be temporary be apart of the object...

Also, how does this handle garbage collection for rooms that get deleted?

bumbadadabum commented:
Hmm yeah good point. Maybe check if every room exists when loading the file?

panpawn commented:
That sounds fair.

I still don't think we should allow it in rooms that are meant to be temporary, though.

bumbadadabum commented:
Hmm yeah good point. Maybe check if every room exists when loading the file?

bumbadadabum commented:
As far as I can see, the trivia plugin also doesn't remove things from a deleted room. I don't see it as a large problem in this case.

panpawn commented:
Although, if you did put this in the room object and limited the amount of FAQ commands a room can have, you wouldn't have to deal with garbage collection at all, and it would also remove a good portion of this code, too.

panpawn commented on chat-plugins/faq.js:
```diff
@@ -0,0 +1,59 @@
+		if (!roomFaqs[room.id]) roomFaqs[room.id] = {};
+		roomFaqs[room.id][topic] = text.join(',').trim();
+		saveRoomFaqs();
+		this.privateModCommand(`(${user.name} has added a faq for '${topic}'.)`);
```
FAQ is an acronym, and it should be capitalized within the UI.

panpawn commented on chat-plugins/faq.js:
```diff
@@ -0,0 +1,59 @@
+		delete roomFaqs[room.id][topic];
+		if (!Object.keys(roomFaqs[room.id]).length) delete roomFaqs[room.id];
+		saveRoomFaqs();
+		this.privateModCommand(`(${user.name} has removed the faq for '${topic}'.)`);
```
Same here

panpawn commented on chat-plugins/faq.js:
```diff
@@ -0,0 +1,59 @@
+		this.privateModCommand(`(${user.name} has removed the faq for '${topic}'.)`);
+	},
+	roomfaq: function (target, room, user) {
+		if (!roomFaqs[room.id]) return this.errorReply("This room has no faq topics.");
```
And here

panpawn commented on chat-plugins/faq.js:
```diff
@@ -0,0 +1,59 @@
+		if (!roomFaqs[room.id][topic]) return this.sendReply("Invalid topic.");
+		return this.sendReply(roomFaqs[room.id][topic]);
+	},
+	roomfaqhelp: ["/roomfaq - Shows the list of all available faq topics",
```
And in the help commands, too

panpawn commented:
I really don't think this is the right way to go about the concept I was talking about in the Staff room for this feature.

bumbadadabum commented:
Then what is? I don't know what the hell you were talking about but it seemed like you were talking about the command I had already implemented in my bot. And that is what I implemented pretty much exactly. If that's now what you had in mind how about you make your own PR.

Morfent commented:
I'd rather avoid mutating the room object if humanly possible, especially since I want to move the database to SQLite at some point and open it up to all rooms. It makes more sense to just prevent groupchats and battles crom running them.

panpawn commented:
Did you intend on purposely not escaping HTML in this?

panpawn commented:
This code is susceptible to constructor crashes.

bumbadadabum commented:
>Did you intend on purposely not escaping HTML in this?
What needs to be escaped then? this.sendReply already escapes it.

>This code is susceptible to constructor crashes. 
How so?

bumbadadabum commented on chat-plugins/faq.js:
```diff
@@ -0,0 +1,59 @@
+		if (!topic) return this.sendReplyBox("List of topics in this room: " + Object.keys(roomFaqs[room.id]).join(', '));
+		if (!this.runBroadcast()) return;
+
+		if (!roomFaqs[room.id][topic]) return this.sendReply("Invalid topic.");
```
That was intentional, but I guess it makes more sense to not broadcast it.

panpawn commented:
> What needs to be escaped then? this.sendReply already escapes it.

No...  A RO could do ``/addfaq test, |raw| html here`` and that would do what you would expect.

>This code is susceptible to constructor crashes.
How so?

``/roomfaq constructor``

panpawn commented:
This actually is a huge security issue by using ``this.sendReply``.

A malicous RO could impersonate someone by doing something like: ``/addfaq worlddomination, |c|~PS Admin|IM GOING TO BAN EVERYONE HAHAHA``

bumbadadabum commented:
I didn't know it allowed it this way. You don't have to fucking lecture me in every fucking comment by the way.

bumbadadabum commented:
It's **REALLY** starting to get on my nerves.

panpawn commented:
I'm just trying to point out the specific security risks this imposes, as you hadn't considered them, I wanted to make sure you were aware of them specifically, sorry for trying to help?

panpawn commented:
I just don't want this to be merged prematurely, and then have these very security risks on the production server, which is why I feel the need to point them out.

bumbadadabum commented:
Yeah I didn't realize it myself, I'm not mad about you pointing this out at all.

trixiemiller commented on chat-plugins/faq.js:
```diff
@@ -0,0 +1,59 @@
+		let [topic, ...text] = target.split(',');
+
+		if (!(topic && text.length)) return this.parse('/help roomfaq');
+		topic = toId(topic);
```
Why is the ``toId(topic)`` done after the check for whether or not there is a topic?   
This would allow people to do ``/addfaq ( ͡° ͜ʖ ͡° ), dropped my bag of lenny faces`` and have it save as ``roomFaqs[room.id][""]``, which is not callable using /roomfaq and not deleteable with /removefaq

bumbadadabum commented:
It should work better now.

bumbadadabum commented on chat-plugins/faq.js:
```diff
@@ -0,0 +1,70 @@
+		if (text.length > 500) return this.errorReply("FAQ entries should not exceed 500 characters.");
+
+		let htmltags = text.toLowerCase().match(/<\/?(\w+)\b/g);
+		if (htmltags && htmltags.some(val => !ALLOWED_HTML.includes(toId(val)))) {
```
Well they can technically bypass it with /eval but I don't want any other tags in here even by admins.

panpawn commented:
Orda-Y just gave me the idea of using ``|uhtml|`` in this for a show all, kind of a thing.

What if it uses ``|uhtml|`` so that each topic has a button for a user to click to hide/show their respected messages?

That way, we wouldn't have to use detail and summary tags to pray on browser support, we can just do it our own way.  And the functionality for this is already a thing, too!

bumbadadabum commented:
I kinda like the way it currently is. The idea with a |uhtml| crossed my mind but I didn't think it'd be ideal for the usecase of this command (the way it's being used in the wifi room right now for example), in which it's used to broadcast a specific topic. If a room wants to have a big faq page, they could also just make a webpage and host is there. For use in a PS chat this way seems to make more sense to me.

panpawn commented:
Well, I think my thinking is that a room /could/ have quite a few FAQ entries.

Also, I wouldn't say that broadcasting them all at once would be a good idea, but maybe doing something like roomsettings does might not be a bad idea.  Just something to consider, I suppose.

Zarel commented:
Who even is @trixiemiller? Don't thumbs-down a pullreq without commenting, that's just rude. :(

panpawn commented:
@Zarel - she did comment

Zarel commented:
Oh I didn't see @_@

Zarel commented on chat-plugins/faq.js:
```diff
@@ -0,0 +1,72 @@
+		if (!this.canTalk()) return this.errorReply("You cannot do this while unable to talk.");
+		if (!this.can('declare', null, room)) return false;
+		if (!room.chatRoomData) return this.errorReply("This command is unavailable in temporary rooms.");
+		let [topic, ...text] = target.split(',');
```
Should maybe support `|` and `\n` as separators, like in Polls?

bumbadadabum replied:
I didn't really see the point for (small) FAQ messages. Was thinking about adding it later if people really wanted it.

AustinXII commented:
@Zarel Trixie is Sparkychild

Zarel commented:
In conclusion:

- This is not really something I'd personally think was necessary
- It'll take up some RAM

On the other hand:

- Apparently people have been wanting this feature
- The RAM use is comparatively negligible

I guess it can stay in.
